src/array/cpu/array_op_impl.cc

```diff
@@ -256,6 +259,53 @@ IdArray NonZero(BoolArray bool_arr) {
 template IdArray NonZero<kDLCPU, int32_t>(BoolArray bool_arr);
 template IdArray NonZero<kDLCPU, int64_t>(BoolArray bool_arr);
 
+///////////////////////////// SetDiff1d /////////////////////////////
+
+template <DLDeviceType XPU, typename IdType>
+IdArray SetDiff1d(IdArray arr1, IdArray arr2) {
+  CHECK(arr1->ndim == 1) << "SetDiff1d only supports 1D array";
+  CHECK(arr2->ndim == 1) << "SetDiff1d only supports 1D array";
+  const IdType* arr1_data = static_cast<IdType*>(arr1->data);
+  const IdType* arr2_data = static_cast<IdType*>(arr2->data);
+  phmap::flat_hash_set<IdType> arr2_set(arr2_data, arr2_data + arr2->shape[0]);
+
+  phmap::flat_hash_set<IdType> ret_set;
+  for (int64_t i = 0; i < arr1->shape[0]; i++) {
+    if (!arr2_set.contains(arr1_data[i])) {
+      ret_set.insert(arr1_data[i]);
+    }
+  }
+
+  IdArray ret_ndarray =
+    NewIdArray(ret_set.size(), DLContext{kDLCPU, 0}, sizeof(IdType) * 8);
+  IdType* ret_data = static_cast<IdType*>(ret_ndarray->data);
+  std::copy(ret_set.begin(), ret_set.end(), ret_data);
+  std::sort(ret_data, ret_data + ret_set.size());
+  return ret_ndarray;
+}
+
+template IdArray SetDiff1d<kDLCPU, int32_t>(IdArray arr1, IdArray arr2);
+template IdArray SetDiff1d<kDLCPU, int64_t>(IdArray arr1, IdArray arr2);
+
+///////////////////////////// Unique /////////////////////////////
+
+template <DLDeviceType XPU, typename IdType>
+IdArray Unique(IdArray arr) {
+  CHECK(arr->ndim == 1) << "Unique only supports 1D array";
+  const IdType* arr_data = static_cast<IdType*>(arr->data);
+  phmap::flat_hash_set<IdType> arr_set(arr_data, arr_data + arr->shape[0]);
+
+  IdArray ret_ndarray =
+    NewIdArray(arr_set.size(), DLContext{kDLCPU, 0}, sizeof(IdType) * 8);
+  IdType* ret_data = static_cast<IdType*>(ret_ndarray->data);
+  std::copy(arr_set.begin(), arr_set.end(), ret_data);
+  std::sort(ret_data, ret_data + arr_set.size());
```

Comment:
why don't you sort the array and get unique values? it'll be faster than using hashtable.

Reply:
Changed to use sort now